test/runner/e2e_read_test.cpp

```diff
@@ -2,160 +2,65 @@
-    runTest(
-        TestHelper::appendKuzuRootPath("test/test_files/tinysnb/projection/multi_query_part.test"));
-    runTest(TestHelper::appendKuzuRootPath("test/test_files/tinysnb/projection/multi_label.test"));
-}
-
-TEST_F(TinySnbReadTest, Subquery) {
-    runTest(TestHelper::appendKuzuRootPath("test/test_files/tinysnb/subquery/exists.test"));
-    runTest(TestHelper::appendKuzuRootPath("test/test_files/tinysnb/subquery/multi_label.test"));
-}
-
-TEST_F(TinySnbReadTest, OptionalMatch) {
-    runTest(TestHelper::appendKuzuRootPath(
-        "test/test_files/tinysnb/optional_match/optional_match.test"));
-    runTest(
-        TestHelper::appendKuzuRootPath("test/test_files/tinysnb/optional_match/multi_label.test"));
-}
-
-TEST_F(TinySnbReadTest, OrderBy) {
-    runTestAndCheckOrder(
-        TestHelper::appendKuzuRootPath("test/test_files/tinysnb/order_by/single_label.test"));
-    runTestAndCheckOrder(
-        TestHelper::appendKuzuRootPath("test/test_files/tinysnb/order_by/multi_label.test"));
-}
-
-TEST_F(TinySnbReadTest, Union) {
-    runTest(TestHelper::appendKuzuRootPath("test/test_files/tinysnb/union/union.test"));
-}
-
-TEST_F(TinySnbReadTest, Unwind) {
-    runTest(TestHelper::appendKuzuRootPath("test/test_files/tinysnb/unwind/unwind.test"));
-}
-
-TEST_F(TinySnbReadTest, VarLengthExtendTests) {
-    runTest(TestHelper::appendKuzuRootPath(
-        "test/test_files/tinysnb/var_length_extend/var_length_adj_list_extend.test"));
-    runTest(TestHelper::appendKuzuRootPath(
-        "test/test_files/tinysnb/var_length_extend/var_length_column_extend.test"));
-}
-
-TEST_F(OneDimNpyReadTest, Match) {
-    runTest(TestHelper::appendKuzuRootPath("test/test_files/npy-1d/match.test"));
-}
-
-class SingleSourceShortestPathTest : public DBTest {
-public:
-    std::string getInputDir() override {
-        return TestHelper::appendKuzuRootPath("dataset/shortest-path-tests/");
+void registerTests(const std::vector<TestConfig> testConfig) {
+    for (auto testItem : testConfig) {
+        testing::RegisterTest(testItem.testGroup.c_str(), testItem.testName.c_str(), nullptr,
+            nullptr, __FILE__, __LINE__,
+            [=]() -> DBTest* { return new EndToEndReadTest(testItem); });
     }
-};
-
-TEST_F(SingleSourceShortestPathTest, BFS_SSSP) {
-    runTest(TestHelper::appendKuzuRootPath("test/test_files/shortest_path/bfs_sssp.test"));
-    runTest(TestHelper::appendKuzuRootPath("test/test_files/shortest_path/bfs_sssp_large.test"));
 }
 
-class TinySnbSerialReadTest : public DBTest {
-public:
-    std::string getInputDir() override {
-        return TestHelper::appendKuzuRootPath("dataset/tinysnb-serial/");
+std::vector<TestConfig> scanTestFiles(const std::string& path) {
+    TestConfig testConfig;
+    std::vector<TestConfig> tests;
+    std::string previousDirectory;
+    for (const auto& entry : std::filesystem::recursive_directory_iterator(path)) {
```

Comment:
I'm not familiar with `recursive_directory_iterator`. Just one question here: I guess `recursive_directory_iterator` have the iteration order guarantee? It seems you rely on that as you are checking `FileUtils::getParentPathStem(entry) != previousDirectory`. Specifically, I just want to make sure that the following case is covered in this function.

```
test_files/test.group
test_files/t1.test
test_files/group1/test.group
test_files/group1/g1.test
test_files/group2/case1/test.group
test_files/group2/case1/g2c1.test
```

Alternatively, I think something like the following also makes sense, if you don't want to rely on the `recursive_directory_iterator`:
```
void scanTestFiles(const std::string& path, std::vector<TestConfig>& configs) {
    for (entry in path) {
        if (test group file exists) { 
            auto config = parseTestGroup(path); 
            if (config.isValid()) {
                configs.push_back(config);
            }
        }
        if (entry is directory) { scanTestFiles(entry.path, configs); }
    }
}

TestConfig parseTestGroup(const std::string& path) {
    auto testConfig = TestHelper::parseGroupFile(testGroupFile);
    for (entry in path) {
        if (entry is a file && entry's extension is ".test") {
            testConfig.files.push_back(entry.path().string());
        }
    }
    return testConfig;
}
```

Your call here. I don't have preferences over these two solutions.

Reply:
Thanks, that's a good point.
From my tests, I haven't seen something like this (which will result in failure):
```
test_files/group2/case1/test.group
test_files/group1/g1.test
test_files/group2/case1/g2c1.test
```
However, looking at the documentation, it seems the order is not guaranteed. 
Your solution is more reliable and clear, I'll implement that.

Reply:
I initially implemented your code, the only necessary change was to add a bool to check if the current directory was processed and not create multiple configs. It worked, but I slightly changed it to check the directories first and then parse them. I believe it's cleaner, the functions are very small. I hope you're ok with that.

Reply:
Sure! LGTM!